service/forward_service.cc

```diff
@@ -0,0 +1,210 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "service/forward_service.hh"
+
+#include <boost/range/algorithm/remove_if.hpp>
+#include <seastar/core/future-util.hh>
+#include <seastar/core/coroutine.hh>
+
+#include "schema_registry.hh"
+#include "database.hh"
+#include "query-ranges-to-vnodes.hh"
+#include "locator/abstract_replication_strategy.hh"
+#include "gms/gossiper.hh"
+#include "log.hh"
+#include "idl/forward_request.dist.hh"
+#include "message/messaging_service.hh"
+
+#include "service/pager/query_pagers.hh"
+#include "cql3/selection/selectable.hh"
+#include "cql3/selection/selectable-expr.hh"
+#include "cql3/selection/selection.hh"
+#include "cql3/selection/raw_selector.hh"
+#include "cql3/cql_config.hh"
+#include "cql3/column_identifier.hh"
+#include "cql3/query_options.hh"
+#include "cql3/result_set.hh"
+#include "cql_serialization_format.hh"
+
+namespace service {
+static constexpr int DEFAULT_INTERNAL_PAGING_SIZE = 10000;
+static logging::logger flogger("forward_service");
+
+static inline
+const dht::token& end_token(const dht::partition_range& r) {
+    static const dht::token max_token = dht::maximum_token();
+    return r.end() ? r.end()->value().token() : max_token;
+}
+
+static inet_address_vector_replica_set get_live_endpoints(keyspace& ks, const dht::token& token) {
+    auto erm = ks.get_effective_replication_map();
+    inet_address_vector_replica_set eps = erm->get_natural_endpoints_without_node_being_replaced(token);
+    auto itend = boost::range::remove_if(eps, std::not1(std::bind1st(std::mem_fn(&gms::gossiper::is_alive), &gms::get_local_gossiper())));
+    eps.erase(itend, eps.end());
+    return eps;
+}
+
+locator::token_metadata_ptr forward_service::get_token_metadata_ptr() const noexcept {
+    return _shared_token_metadata.get();
+}
+
+future<> forward_service::stop() {
+    return make_ready_future<>();
+}
+
+// Due to `cql3::selection::selection` not being serializable, it cannot be
+// stored in `forward_request`. It has to mocked on the receiving node,
+// based on requested reduction types.
+static shared_ptr<cql3::selection::selection> mock_selection(const std::vector<query::forward_request::reduction_type>& reduction_types, schema_ptr schema, database& db) {
+    std::vector<shared_ptr<cql3::selection::raw_selector>> raw_selectors;
+
+    auto mock_singular_selection = [&] (const query::forward_request::reduction_type& type) {
+        switch (type) {
+            case query::forward_request::reduction_type::count: {
+                auto selectable = cql3::selection::make_count_rows_function_expression();
+                auto column_identifier = make_shared<cql3::column_identifier>("count", false);
+                return make_shared<cql3::selection::raw_selector>(selectable, column_identifier);
+            }
+            default:
```

Comment:
Maybe it's better to do this in select_statement.cc to avoid cql3/ spreading over more code. forwarding_service should have a cql3-neutral way to represent the reduction.

Reply:
We can consider moving cql3/expr out of cql3 if it helps.

Reply:
I thought of it in such a way that the `forwarding_service` already has a cql3-neutral way to represent a reduction - reduction_type enum. 
Also, I don't think that moving code responsible for `cql3::selection::selection` mocking would help significantly in reducing `cql3/` spread. `cql3` is still needed here for `result_set`, `result_set_builder` and pager creation. Current solution is quite hacky, but I'm in favor of keeping hacky things together - in `forward_service`.

Reply:
@avikivity ping